### PR TITLE
Pakwheels scraper

### PR DESCRIPTION
**Pull request by A-ASAD**
Scraps Name, Price, and Description of newly launched cars from Pakwheels.com.

### CODE REVIEW

File: pakwheels/pakwheels/spiders/github_spider.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            url='https://github.com/session',
+            method="POST",
+            formdata={
+                'authenticity_token': authenticity_token,
+
+                # your login here (email/username)
+                'login': '',
+
+                # your password here
+                'password': ''
```
Comment (saadyousafarbi) [inline pakwheels/pakwheels/spiders/github_spider.py:40]: Get these from environment variables!

File: pakwheels/pakwheels/spiders/github_spider.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        name = response.css('.p-name::text').get().strip()
+        nickname = response.css('.p-nickname::text').get().strip()
+        followers, following, stars = response.css(
+            'span.text-bold.color-text-primary::text'
+            ).getall()
+        repositories = response.css(
+            '#user-repositories-list a[itemprop="name codeRepository"]::text'
+            ).getall()
+        repositories = ', '.join(map(str.strip, repositories))
```
Comment (saadyousafarbi) [inline pakwheels/pakwheels/spiders/github_spider.py:83]: use xpath selectors for the github crawler

File: pakwheels/pakwheels/spiders/pakwheels_spider.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            'https://www.pakwheels.com/used-cars/',
+        ]
+        for url in urls:
+            if url == 'https://www.pakwheels.com/new-cars/':
+                yield scrapy.Request(
+                    url=url,
+                    callback=self.parse_newly_launched_cars
+                    )
+            else:
+                yield scrapy.Request(url=url, callback=self.parse_used_cars)
```
Comment (saadyousafarbi) [inline pakwheels/pakwheels/spiders/pakwheels_spider.py:23]: you should have the same callback method for handling response. In the callback method, you can have a different method for used and new cars!

File: pakwheels/pakwheels/spiders/github_spider.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        Navigates to repositories page
+
+        Arguments:
+            self
+            response:
+                Response object
+        """
+
+        yield response.follow(
+            'https://github.com/A-ASAD?tab=repositories',
```
Comment (saadyousafarbi) [inline pakwheels/pakwheels/spiders/github_spider.py:57]: the username here needs to come from the username, should not be hardcoded!

**Review by saadyousafarbi — CHANGES_REQUESTED** [review]
There are some comments I have added, please have a look!

Please write the results to a CSV file, instead of only printing them out.

File: pakwheels/pakwheels/spiders/github_spider.py (hunk 4 above)
Comment (A-ASAD) [reply pakwheels/pakwheels/spiders/github_spider.py:59]: I've explored all the page and retraced javascript requests, I'm unable to get the username from the dashboard. So, I'll have to hard code it here.

File: pakwheels/pakwheels/middlewares.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+class PakwheelsSpiderMiddleware:
+    # Not all methods need to be defined. If a method is not defined,
+    # scrapy acts as if the spider middleware does not modify the
+    # passed objects.
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        # This method is used by Scrapy to create your spiders.
+        s = cls()
```
Comment (AhtishamShahid) [inline pakwheels/pakwheels/middlewares.py:20]: Use a more descriptive name.

File: pakwheels/pakwheels/middlewares.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+        # Should return None or raise an exception.
+        return None
+
+    def process_spider_output(self, response, result, spider):
+        # Called with the results returned from the Spider, after
+        # it has processed the response.
+
+        # Must return an iterable of Request, or item objects.
+        for i in result:
```
Comment (AhtishamShahid) [inline pakwheels/pakwheels/middlewares.py:36]: same here , use more descriptive naming.

File: pakwheels/pakwheels/spiders/github_spider.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            self
+            response:
+                Response object
+
+        Yields:
+            User data (name, nickname, followers, following, stars,
+            repositories)
+        """
+
+        name = response.xpath(
```
Comment (AhtishamShahid) [inline pakwheels/pakwheels/spiders/github_spider.py:77]: These can be moved in different functions.

File: pakwheels/pakwheels/middlewares.py (hunk 5 above)
Comment (A-ASAD) [reply pakwheels/pakwheels/middlewares.py:20]: This is an auto-generated file having some default implementation for middleware.

File: pakwheels/pakwheels/middlewares.py (hunk 6 above)
Comment (A-ASAD) [reply pakwheels/pakwheels/middlewares.py:36]: This is an auto-generated file having some default implementation for middleware.

**Pull request merged**